chain/index/chain_indexer.go

```diff
@@ -0,0 +1,723 @@
+	tsKeyCid, err := ts.Key().Cid()
+	if err != nil {
+		return nil, xerrors.Errorf("error getting tipset key cid: %w", err)
+	}
+	return tsKeyCid.Bytes(), nil
+}
+
+func isIndexedValue(b uint8) bool {
+	// currently we mark the full entry as indexed if either the key
+	// or the value are indexed; in the future we will need finer-grained
+	// management of indices
+	return b&(types.EventFlagIndexedKey|types.EventFlagIndexedValue) > 0
+}
+
+func ethTxHashFromSignedMessage(smsg *types.SignedMessage) (ethtypes.EthHash, error) {
+	if smsg.Signature.Type == crypto.SigTypeDelegated {
+		tx, err := ethtypes.EthTransactionFromSignedFilecoinMessage(smsg)
+		if err != nil {
+			return ethtypes.EthHash{}, xerrors.Errorf("failed to convert from signed message: %w", err)
+		}
+
+		return tx.TxHash()
+	} else if smsg.Signature.Type == crypto.SigTypeSecp256k1 {
+		return ethtypes.EthHashFromCid(smsg.Cid())
+	}
+	// else BLS message
+	return ethtypes.EthHashFromCid(smsg.Message.Cid())
+}
+
+type executedMessage struct {
+	msg types.ChainMsg
+	rct *types.MessageReceipt
+	// events extracted from receipt
+	evs []*types.Event
+}
+
+func (ci *ChainIndexer) loadExecutedMessages(ctx context.Context, msgTs, rctTs *types.TipSet) ([]executedMessage, error) {
+	msgs, err := ci.cs.MessagesForTipset(ctx, msgTs)
+	if err != nil {
+		return nil, xerrors.Errorf("read messages: %w", err)
+	}
+
+	st := ci.cs.ActorStore(ctx)
+
+	arr, err := blockadt.AsArray(st, rctTs.Blocks()[0].ParentMessageReceipts)
+	if err != nil {
+		return nil, xerrors.Errorf("load receipts amt: %w", err)
+	}
+
+	if uint64(len(msgs)) != arr.Length() {
+		return nil, xerrors.Errorf("mismatching message and receipt counts (%d msgs, %d rcts)", len(msgs), arr.Length())
+	}
+
+	ems := make([]executedMessage, len(msgs))
+
+	for i := 0; i < len(msgs); i++ {
+		ems[i].msg = msgs[i]
+
+		var rct types.MessageReceipt
+		found, err := arr.Get(uint64(i), &rct)
+		if err != nil {
+			return nil, xerrors.Errorf("load receipt: %w", err)
+		}
+		if !found {
+			return nil, xerrors.Errorf("receipt %d not found", i)
+		}
+		ems[i].rct = &rct
+
+		if rct.EventsRoot == nil {
+			continue
+		}
+
+		evtArr, err := amt4.LoadAMT(ctx, st, *rct.EventsRoot, amt4.UseTreeBitWidth(types.EventAMTBitwidth))
+		if err != nil {
+			return nil, xerrors.Errorf("load events amt: %w", err)
+		}
+
+		ems[i].evs = make([]*types.Event, evtArr.Len())
+		var evt types.Event
+		err = evtArr.ForEach(ctx, func(u uint64, deferred *cbg.Deferred) error {
+			if u > math.MaxInt {
+				return xerrors.Errorf("too many events")
+			}
+			if err := evt.UnmarshalCBOR(bytes.NewReader(deferred.Raw)); err != nil {
+				return err
+			}
+
+			cpy := evt
+			ems[i].evs[int(u)] = &cpy //nolint:scopelint
+			return nil
+		})
+
+		if err != nil {
+			return nil, xerrors.Errorf("read events: %w", err)
+		}
+
+	}
+
+	return ems, nil
+}
+
+func (ci *ChainIndexer) GetMsgInfo(ctx context.Context, msg_cid cid.Cid) (MsgInfo, error) {
```

Comment:
hopefully the linter is going to have issues with your underscore in `msg_cid`

Reply:
Sorry, trying out some new tooling. Will fix all this.